File: OctoberChallenge/week2/house-robber-2.java
class Solution {
    public int rob(int[] nums) {
        // Early return: Edge Cases
        if(nums.length == 0) return 0;
        if(nums.length == 1) return nums[0];
        if(nums.length == 2) return Math.max(nums[0], nums[1]);
        
        // Final case
        int[] p = Arrays.copyOfRange(nums, 0, nums.length-1);
        int[] q = Arrays.copyOfRange(nums, 1, nums.length);
        return Math.max(houseRobberI(p), houseRobberI(q));
    }

    int houseRobberI(int[] nums) {
        int[] res = new int[nums.length+1];
        res[0] = nums[0];
        res[1] = Math.max(nums[0], nums[1]);
        for(int k=2; k < nums.length; k++) {
            res[k] = Math.max(nums[k]+res[k-2], res[k-1]);
        }
        return res[nums.length-1];
    }
}

File: OctoberChallenge/week2/house-robber-2.cpp
class Solution {
public:
    int rob(vector<int>& nums) {
        // Early return: Edge Cases
        if(nums.size() == 0) return 0;
        if(nums.size() == 1) return nums[0];
        if(nums.size() == 2) return max(nums[0], nums[1]);
        
        // Final case
        vector<int> p(nums.begin(), nums.end()-1);
        vector<int> q(nums.begin()+1, nums.end());
        return max(houseRobberI(p), houseRobberI(q));
    }
    
    int houseRobberI(vector<int>& nums) {
        int res[nums.size()+1];
        res[0] = nums[0];
        res[1] = max(nums[0], nums[1]);
        for(int k=2; k < nums.size(); k++) res[k] = max(nums[k]+res[k-2], res[k-1]);
        return res[nums.size()-1];
    }
};
